src/storage/checkpoint_manager.cpp

```diff
@@ -0,0 +1,232 @@
+#include "storage/checkpoint_manager.h"
+#include <unordered_set>
+#include <vector>
+#include "common/macros.h"
+
+#define NUM_RESERVED_COLUMNS 1u
+
+namespace terrier::storage {
+void CheckpointManager::Checkpoint(const SqlTable &table, const catalog::Schema &schema, uint32_t catalog_type) {
+  std::vector<catalog::col_oid_t> all_col(schema.GetColumns().size());
+  uint16_t col_idx = 0;
+  for (const catalog::Schema::Column &column : schema.GetColumns()) {
+    all_col[col_idx] = column.GetOid();
+    col_idx++;
+  }
+
+  out_.SetTableOid(table.Oid());
+  if (catalog_type == 0) {  // normal table
+    // TODO(mengyang): should be the version of the table.
+    out_.SetVersion(0);
+  } else {  // catalog table
+    out_.SetVersion(catalog_type);
+  }
+
+  auto row_pair = table.InitializerForProjectedRow(all_col);
+  auto *buffer = common::AllocationUtil::AllocateAligned(row_pair.first.ProjectedRowSize());
+  ProjectedRow *row_buffer = row_pair.first.InitializeRow(buffer);
+
+  for (auto &slot : table) {
+    if (table.Select(txn_, slot, row_buffer)) {
+      out_.SerializeTuple(row_buffer, &slot, schema, row_pair.second);
+    }
+  }
+  out_.Persist();
+  delete[] buffer;
+}
+
+void CheckpointManager::Recover(const char *checkpoint_file_path) {
+  BufferedTupleReader reader(checkpoint_file_path);
+
+  while (reader.ReadNextBlock()) {
+    CheckpointFilePage *page = reader.GetPage();
+    // TODO(zhaozhe): check checksum here
+    catalog::table_oid_t oid = page->GetTableOid();
+    SqlTable *table = GetTable(oid);
+
+    ProjectedRow *row = nullptr;
+    while ((row = reader.ReadNextRow()) != nullptr) {
+      TupleSlot *slot UNUSED_ATTRIBUTE = reader.ReadNextTupleSlot();
+      // loop through columns to deal with non-inlined varlens.
+      for (uint16_t projection_list_idx = 0; projection_list_idx < row->NumColumns(); projection_list_idx++) {
+        if (!row->IsNull(projection_list_idx)) {
+          catalog::col_oid_t col = table->ColOidForId(row->ColumnIds()[projection_list_idx]);
+          if (table->GetSchema().GetColumn(col).IsVarlen()) {
```

Comment:
The point is that even if we recover the catalog first, we wouldn't be able to get the schema, because the catalog is not ready. It lacks fundamental APIs.

Reply:
According to the project meeting, we plan to implement catalog recovery in this way:

Start DBMS for the first time:
1. Create catalog tables.
2. Fill catalog tables with initial content.
3. Initialize catalog data structure according to contents of catalog tables.

Recovering DBMS:
1. Create catalog tables.
2. Recovery code fills the catalog tables with contents in the checkpoint.
3. Recovery code notifies catalog to initialize its data structure according to contents of catalog tables (including creating user tables).
4. Recovery code fills the user tables with contents in the checkpoint.
5. Recovery code recovers the log. During the process, we may encounter records that modify catalog tables. We will notify catalog to update correspondingly after redoing them.

Currently, the catalog does not meet our requirement:
1) It always fills catalog tables with initial content when creating catalog tables.
2) No API to notify that we have updated catalog tables and catalog needs to update correspondingly.